core/os/android/profiling/cc/launch_producer.cc

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <android/log.h>
+#include <dlfcn.h>
+#include <string>
+#include <string.h>
+#include <sys/system_properties.h>
+
+#define _LOG(lvl, name, msg, ...)                           \
+  do {                                                      \
+    printf(name ": " msg "\n", ## __VA_ARGS__);             \
+    __android_log_print(lvl, "GAPID", msg, ## __VA_ARGS__); \
+  } while (false)
+
+#define LOG_ERR(msg, ...)  _LOG(ANDROID_LOG_ERROR, "E", msg, ## __VA_ARGS__)
+#define LOG_WARN(msg, ...) _LOG(ANDROID_LOG_WARN,  "W", msg, ## __VA_ARGS__)
+#define LOG_INFO(msg, ...) _LOG(ANDROID_LOG_INFO,  "I", msg, ## __VA_ARGS__)
+
+#define GET_PROP(name, trans)                       \
+  do {                                              \
+    char _v[PROP_VALUE_MAX] = {0};                  \
+    if (__system_property_get(name, _v) == 0) {     \
+      LOG_ERR("Failed reading property %s", name);  \
+      std::abort();                                 \
+    }                                               \
+    trans;                                          \
+  } while (0)
+#define GET_STRING_PROP(n, t) GET_PROP(n, t = _v)
+
+typedef void (*FN_PTR)(void);
+
+const char* kDriverProperty = "ro.gfx.driver.1";
+const char* kProducerSO = "/assets/libgpudataproducer.so";
```

Comment:
We can't hardcode this, eventually this .so file is in lib/{ABI}/libgpudataproducer.so. Can a binary tell the ABI of itself?

Reply:
Done.